File: medbillpro/src/core/services/importer.js
'use strict';

const db = require('../db');
const products = require('./products');
const settings = require('./settings');
const { readAny, serialToIso } = require('../util/xlsx');
const { assert } = require('../util/errors');
const { r2, r3 } = require('../util/money');
const { isExpiryMonth } = require('../util/dates');

/**
 * Purchase-bill import. Distributor sheets never share a layout, so columns
 * are detected from the header row by keyword, and anything ambiguous comes
 * back to the screen as an editable preview instead of being guessed at.
 */
const FIELDS = {
  name: ['product', 'item', 'medicine', 'description', 'particulars', 'name', 'product name', 'item name'],
  pack: ['pack', 'packing'],
  batch: ['batch', 'batch no', 'batchno', 'b.no', 'bno', 'lot'],
  expiry: ['expiry', 'exp', 'exp date', 'expdt', 'exp.dt', 'expiry date'],
  qty: ['qty', 'quantity', 'qnty', 'nos', 'units', 'pcs'],
  free: ['free', 'free qty', 'scheme', 'bonus'],
  mrp: ['mrp', 'm.r.p', 'mrp rs', 'retail'],
  rate: ['rate', 'ptr', 'purchase rate', 'price', 'net rate', 'basic', 'trade rate'],
  disc: ['disc', 'discount', 'disc %', 'dis%', 'cd'],
  gst: ['gst', 'gst %', 'tax', 'igst', 'cgst+sgst', 'gst rate', 'vat'],
  hsn: ['hsn', 'hsn code'],
  mfr: ['mfr', 'manufacturer', 'company', 'mfg', 'marketer']
};

function normaliseHeader(cell) {
  return String(cell || '').toLowerCase().replace(/[._]/g, ' ').replace(/\s+/g, ' ').trim();
}

function detectColumns(headerRow) {
  const map = {};
  headerRow.forEach((cell, index) => {
    const h = normaliseHeader(cell);
    if (!h) return;
    for (const [field, keys] of Object.entries(FIELDS)) {
      if (map[field] !== undefined) continue;
      if (keys.some((k) => h === k)) { map[field] = index; return; }
    }
    for (const [field, keys] of Object.entries(FIELDS)) {
      if (map[field] !== undefined) continue;
      if (keys.some((k) => h.includes(k))) { map[field] = index; return; }
    }
  });
  return map;
}

/** Picks the header row — the first row that names a product and a quantity. */
function findHeaderRow(rows) {
  for (let i = 0; i < Math.min(rows.length, 25); i += 1) {
    const map = detectColumns(rows[i]);
    if (map.name !== undefined && (map.qty !== undefined || map.rate !== undefined)) return { index: i, map };
  }
  return { index: -1, map: {} };
}

function toNumber(value) {
  if (value === undefined || value === null || value === '') return 0;
  const cleaned = String(value).replace(/[₹,\s]/g, '').replace(/[^0-9.+-]/g, '');
  const n = Number(cleaned);
  return Number.isFinite(n) ? n : 0;
}

function parseExpiry(value) {
  const raw = String(value ?? '').trim();
  if (!raw) return null;
  if (isExpiryMonth(raw)) return raw;
  if (/^\d{4}-\d{2}-\d{2}$/.test(raw)) return raw.slice(0, 7);
  const serial = Number(raw);
  if (Number.isFinite(serial) && serial > 1000) {
    const iso = serialToIso(serial);
    if (iso) return iso.slice(0, 7);
  }
  const mmYY = raw.match(/^(\d{1,2})\s*[/\-.]\s*(\d{2}|\d{4})$/);
  if (mmYY) {
    const mm = String(Math.min(12, Math.max(1, Number(mmYY[1])))).padStart(2, '0');
    const yy = mmYY[2].length === 2 ? `20${mmYY[2]}` : mmYY[2];
    return `${yy}-${mm}`;
  }
  const monthName = raw.match(/^([A-Za-z]{3,9})[\s-]*(\d{2}|\d{4})$/);
  if (monthName) {
    const months = ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec'];
    const idx = months.indexOf(monthName[1].slice(0, 3).toLowerCase());
    if (idx >= 0) {
      const yy = monthName[2].length === 2 ? `20${monthName[2]}` : monthName[2];
      return `${yy}-${String(idx + 1).padStart(2, '0')}`;
    }
  }
  const dmy = raw.match(/^(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})$/);
  if (dmy) {
    const mm = String(Math.min(12, Math.max(1, Number(dmy[2])))).padStart(2, '0');
    const yy = dmy[3].length === 2 ? `20${dmy[3]}` : dmy[3];
    return `${yy}-${mm}`;
  }
  return null;
}

/** Units per pack from strings like "10x10", "1x15 TAB", "30 ML". */
function parsePack(value) {
  const raw = String(value ?? '').toLowerCase();
  const cross = raw.match(/(\d+)\s*[x*]\s*(\d+)/);
  if (cross) return Math.max(1, Number(cross[2]));
  const plain = raw.match(/(\d+)/);
  return plain ? Math.max(1, Number(plain[1])) : 1;
}

function matchProduct(name) {
  const conn = db.get();
  const clean = String(name || '').trim();
  if (!clean) return null;
  const exact = conn.prepare('SELECT * FROM products WHERE lower(name) = lower(?) AND active = 1').get(clean);
  if (exact) return exact;
  const like = conn.prepare('SELECT * FROM products WHERE name LIKE ? AND active = 1 ORDER BY length(name) LIMIT 1')
    .get(`${clean.split(/\s+/)[0]}%`);
  if (like && like.name.toLowerCase().startsWith(clean.slice(0, 4).toLowerCase())) return like;
  return null;
}

/**
 * Parses a spreadsheet into draft purchase lines. Nothing is written to the
 * database here — the operator reviews the preview and then saves.
 */
function preview({ filename = '', base64 = '', text = '' }) {
  const buffer = base64 ? Buffer.from(base64, 'base64') : Buffer.from(String(text), 'utf8');
  assert(buffer.length > 0, 'That file is empty.');
  assert(buffer.length < 12 * 1024 * 1024, 'That file is larger than 12 MB — please split it.');
  const rows = readAny(buffer, filename);
  assert(rows.length > 1, 'No rows could be read from that file.');
  const { index, map } = findHeaderRow(rows);
  assert(index >= 0, 'Could not find a header row. The sheet needs column titles such as Product, Batch, Expiry, Qty, Rate, MRP.');

  const defaultGst = settings.num('default_gst_rate', 12);
  const body = rows.slice(index + 1);
  const lines = [];
  for (const row of body) {
    const cell = (field) => (map[field] === undefined ? '' : row[map[field]] ?? '');
    const name = String(cell('name')).trim();
    if (!name) continue;
    if (/^(total|grand total|sub total|net amount|amount)/i.test(name)) continue;
    const qty = Math.round(toNumber(cell('qty')));
    const rate = toNumber(cell('rate'));
    if (qty <= 0 && rate <= 0) continue;
    const packSize = map.pack !== undefined ? parsePack(cell('pack')) : 1;
    const matched = matchProduct(name);
    const expiry = parseExpiry(cell('expiry'));
    const gstRaw = toNumber(cell('gst'));
    const mrp = toNumber(cell('mrp'));
    const issues = [];
    if (!matched) issues.push('New medicine — will be created on import');
    if (!expiry) issues.push('Expiry missing or unreadable');
    if (qty <= 0) issues.push('Quantity missing');
    if (rate <= 0) issues.push('Rate missing');
    lines.push({
      source_name: name,
      product_id: matched?.id ?? null,
      product_name: matched?.name ?? name,
      manufacturer: String(cell('mfr')).trim(),
      hsn: String(cell('hsn')).trim() || '3004',
      pack_size: matched?.pack_size ?? packSize,
      batch_no: String(cell('batch')).trim().toUpperCase() || 'NA',
      expiry,
      qty_units: qty > 0 ? qty * (matched ? 1 : 1) : 0,
      free_units: Math.max(0, Math.round(toNumber(cell('free')))),
      mrp: r2(mrp),
      rate_per_unit: r3(rate),
      disc_pct: Math.min(100, Math.max(0, toNumber(cell('disc')))),
      gst_rate: [0, 5, 12, 18, 28].includes(gstRaw) ? gstRaw : (matched?.gst_rate ?? defaultGst),
      issues
    });
  }
  assert(lines.length > 0, 'No usable item rows were found in that file.');
  return {
    filename,
    headerRow: index + 1,
    columns: map,
    detected: Object.keys(map),
    lines,
    summary: {
      rows: lines.length,
      newProducts: lines.filter((l) => !l.product_id).length,
      withIssues: lines.filter((l) => l.issues.length > 0).length,
      value: r2(lines.reduce((s, l) => s + l.qty_units * l.rate_per_unit, 0))
    }
  };
}

/** Creates any medicines the sheet introduced, returning lines ready to save. */
function materialise(actor, { lines = [] }) {
  assert(Array.isArray(lines) && lines.length, 'Nothing to import.');
  return db.tx(() => lines.map((line) => {
    let productId = line.product_id ? Number(line.product_id) : null;
    if (!productId) {
      const created = products.save(actor, {
        name: line.product_name || line.source_name,
        manufacturer: line.manufacturer || '',
        hsn: line.hsn || '3004',
        gst_rate: line.gst_rate,
        pack_size: line.pack_size || 1,
        category: 'Medicine'
      });
      productId = created.id;
    }
    return {
      product_id: productId,
      batch_no: line.batch_no,
      expiry: line.expiry,
      mrp: line.mrp,
      qty_units: line.qty_units,
      free_units: line.free_units,
      rate_per_unit: line.rate_per_unit,
      disc_pct: line.disc_pct,
      gst_rate: line.gst_rate
    };
  }));
}

/** Medicine-master import — used to load an opening catalogue quickly. */
function importProducts(actor, { filename = '', base64 = '', text = '' }) {
  const buffer = base64 ? Buffer.from(base64, 'base64') : Buffer.from(String(text), 'utf8');
  const rows = readAny(buffer, filename);
  const { index, map } = findHeaderRow(rows);
  assert(index >= 0, 'Could not find a header row with a product name column.');
  const created = [];
  const skipped = [];
  db.tx(() => {
    for (const row of rows.slice(index + 1)) {
      const cell = (f) => (map[f] === undefined ? '' : row[map[f]] ?? '');
      const name = String(cell('name')).trim();
      if (!name) continue;
      try {
        created.push(products.save(actor, {
          name,
          manufacturer: String(cell('mfr')).trim(),
          hsn: String(cell('hsn')).trim() || '3004',
          gst_rate: [0, 5, 12, 18, 28].includes(toNumber(cell('gst'))) ? toNumber(cell('gst')) : settings.num('default_gst_rate', 12),
          pack_size: map.pack !== undefined ? parsePack(cell('pack')) : 1
        }).name);
      } catch (err) {
        skipped.push({ name, reason: err.message });
      }
    }
  });
  return { created: created.length, skipped, names: created.slice(0, 50) };
}

module.exports = { preview, materialise, importProducts, detectColumns, parseExpiry, parsePack, toNumber };

File: medv-python/medv/services/importer.py
"""
Purchase-bill import.

Distributor sheets never share a layout, so columns are detected from the header
row by keyword and anything ambiguous comes back to the screen as an editable
preview instead of being guessed at.
"""

from __future__ import annotations
import base64
import re

from .. import db
from ..util.dates import is_expiry_month
from ..util.errors import assert_that
from ..util.money import r2, r3
from ..util.sheets import read_any, serial_to_iso
from . import products, settings

FIELDS = {
    "name": ["product", "item", "medicine", "description", "particulars", "name", "product name", "item name"],
    "pack": ["pack", "packing"],
    "batch": ["batch", "batch no", "batchno", "b.no", "bno", "lot"],
    "expiry": ["expiry", "exp", "exp date", "expdt", "exp.dt", "expiry date"],
    "qty": ["qty", "quantity", "qnty", "nos", "units", "pcs"],
    "free": ["free", "free qty", "scheme", "bonus"],
    "mrp": ["mrp", "m.r.p", "mrp rs", "retail"],
    "rate": ["rate", "ptr", "purchase rate", "price", "net rate", "basic", "trade rate"],
    "disc": ["disc", "discount", "disc %", "dis%", "cd"],
    "gst": ["gst", "gst %", "tax", "igst", "cgst+sgst", "gst rate", "vat"],
    "hsn": ["hsn", "hsn code"],
    "mfr": ["mfr", "manufacturer", "company", "mfg", "marketer"],
}
GST_RATES = (0, 5, 12, 18, 28)
MONTHS = ["jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"]


def _normalise_header(cell) -> str:
    return re.sub(r"\s+", " ", str(cell or "").lower().replace(".", " ").replace("_", " ")).strip()


def detect_columns(header_row) -> dict:
    mapping: dict[str, int] = {}
    headers = [(index, _normalise_header(cell)) for index, cell in enumerate(header_row)]
    for index, head in headers:
        if not head:
            continue
        for field, keys in FIELDS.items():
            if field not in mapping and head in keys:
                mapping[field] = index
                break
        else:
            for field, keys in FIELDS.items():
                if field not in mapping and any(key in head for key in keys):
                    mapping[field] = index
                    break
    return mapping


def find_header_row(rows):
    """The header is the first row that names a product and a quantity or rate."""
    for index, row in enumerate(rows[:25]):
        mapping = detect_columns(row)
        if "name" in mapping and ("qty" in mapping or "rate" in mapping):
            return index, mapping
    return -1, {}


def to_number(value) -> float:
    if value in (None, ""):
        return 0.0
    cleaned = re.sub(r"[^0-9.+-]", "", str(value).replace("₹", "").replace(",", ""))
    try:
        return float(cleaned)
    except ValueError:
        return 0.0


def parse_expiry(value):
    raw = str(value or "").strip()
    if not raw:
        return None
    if is_expiry_month(raw):
        return raw
    if re.match(r"^\d{4}-\d{2}-\d{2}$", raw):
        return raw[:7]
    try:
        serial = float(raw)
        if serial > 1000:
            iso = serial_to_iso(serial)
            if iso:
                return iso[:7]
    except ValueError:
        pass
    match = re.match(r"^(\d{1,2})\s*[/\-.]\s*(\d{2}|\d{4})$", raw)
    if match:
        month = str(min(12, max(1, int(match.group(1))))).zfill(2)
        year = f"20{match.group(2)}" if len(match.group(2)) == 2 else match.group(2)
        return f"{year}-{month}"
    match = re.match(r"^([A-Za-z]{3,9})[\s-]*(\d{2}|\d{4})$", raw)
    if match and match.group(1)[:3].lower() in MONTHS:
        month = str(MONTHS.index(match.group(1)[:3].lower()) + 1).zfill(2)
        year = f"20{match.group(2)}" if len(match.group(2)) == 2 else match.group(2)
        return f"{year}-{month}"
    match = re.match(r"^(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})$", raw)
    if match:
        month = str(min(12, max(1, int(match.group(2))))).zfill(2)
        year = f"20{match.group(3)}" if len(match.group(3)) == 2 else match.group(3)
        return f"{year}-{month}"
    return None


def parse_pack(value) -> int:
    raw = str(value or "").lower()
    cross = re.search(r"(\d+)\s*[x*]\s*(\d+)", raw)
    if cross:
        return max(1, int(cross.group(2)))
    plain = re.search(r"(\d+)", raw)
    return max(1, int(plain.group(1))) if plain else 1


def match_product(name):
    clean = str(name or "").strip()
    if not clean:
        return None
    exact = db.one_row("SELECT * FROM products WHERE lower(name) = lower(?) AND active = 1", (clean,))
    if exact:
        return exact
    first_word = clean.split()[0] if clean.split() else clean
    like = db.one_row(
        "SELECT * FROM products WHERE name LIKE ? AND active = 1 ORDER BY length(name) LIMIT 1",
        (f"{first_word}%",))
    if like and like["name"].lower().startswith(clean[:4].lower()):
        return like
    return None


def _decode(base64_text="", text="") -> bytes:
    if base64_text:
        return base64.b64decode(base64_text)
    return str(text).encode("utf-8")


def preview(filename="", base64_data="", text="", **kwargs):
    """Parses a sheet into draft purchase lines. Nothing is written yet."""
    data = _decode(base64_data or kwargs.get("base64", ""), text)
    assert_that(len(data) > 0, "That file is empty.")
    assert_that(len(data) < 12 * 1024 * 1024, "That file is larger than 12 MB — please split it.")
    rows = read_any(data, filename)
    assert_that(len(rows) > 1, "No rows could be read from that file.")
    index, mapping = find_header_row(rows)
    assert_that(index >= 0,
                "Could not find a header row. The sheet needs column titles such as Product, "
                "Batch, Expiry, Qty, Rate, MRP.")

    default_gst = settings.num("default_gst_rate", 12)
    lines = []
    for row in rows[index + 1:]:
        def cell(field):
            position = mapping.get(field)
            if position is None or position >= len(row):
                return ""
            return row[position]

        name = str(cell("name")).strip()
        if not name or re.match(r"^(total|grand total|sub total|net amount|amount)", name, re.I):
            continue
        qty = round(to_number(cell("qty")))
        rate = to_number(cell("rate"))
        if qty <= 0 and rate <= 0:
            continue
        matched = match_product(name)
        expiry = parse_expiry(cell("expiry"))
        gst_raw = to_number(cell("gst"))
        issues = []
        if not matched:
            issues.append("New medicine — will be created on import")
        if not expiry:
            issues.append("Expiry missing or unreadable")
        if qty <= 0:
            issues.append("Quantity missing")
        if rate <= 0:
            issues.append("Rate missing")
        lines.append({
            "source_name": name,
            "product_id": matched["id"] if matched else None,
            "product_name": matched["name"] if matched else name,
            "manufacturer": str(cell("mfr")).strip(),
            "hsn": str(cell("hsn")).strip() or "3004",
            "pack_size": matched["pack_size"] if matched else (parse_pack(cell("pack")) if "pack" in mapping else 1),
            "batch_no": str(cell("batch")).strip().upper() or "NA",
            "expiry": expiry,
            "qty_units": max(0, qty),
            "free_units": max(0, round(to_number(cell("free")))),
            "mrp": r2(to_number(cell("mrp"))),
            "rate_per_unit": r3(rate),
            "disc_pct": min(100.0, max(0.0, to_number(cell("disc")))),
            "gst_rate": gst_raw if int(gst_raw) in GST_RATES else (matched["gst_rate"] if matched else default_gst),
            "issues": issues,
        })
    assert_that(lines, "No usable item rows were found in that file.")
    return {
        "filename": filename,
        "headerRow": index + 1,
        "columns": mapping,
        "detected": list(mapping),
        "lines": lines,
        "summary": {
            "rows": len(lines),
            "newProducts": sum(1 for line in lines if not line["product_id"]),
            "withIssues": sum(1 for line in lines if line["issues"]),
            "value": r2(sum(line["qty_units"] * line["rate_per_unit"] for line in lines)),
        },
    }


def materialise(actor, payload):
    """Creates any medicine the sheet introduced, returning lines ready to save."""
    lines = payload.get("lines") or []
    assert_that(isinstance(lines, list) and lines, "Nothing to import.")
    with db.tx():
        out = []
        for line in lines:
            product_id = int(line["product_id"]) if line.get("product_id") else None
            if not product_id:
                created = products.save(actor, {
                    "name": line.get("product_name") or line.get("source_name"),
                    "manufacturer": line.get("manufacturer", ""),
                    "hsn": line.get("hsn") or "3004",
                    "gst_rate": line.get("gst_rate"),
                    "pack_size": line.get("pack_size") or 1,
                    "category": "Medicine",
                })
                product_id = created["id"]
            out.append({
                "product_id": product_id,
                "batch_no": line.get("batch_no"),
                "expiry": line.get("expiry"),
                "mrp": line.get("mrp"),
                "qty_units": line.get("qty_units"),
                "free_units": line.get("free_units"),
                "rate_per_unit": line.get("rate_per_unit"),
                "disc_pct": line.get("disc_pct"),
                "gst_rate": line.get("gst_rate"),
            })
        return out


def import_products(actor, payload):
    """Medicine-master import — used to load an opening catalogue quickly."""
    data = _decode(payload.get("base64", ""), payload.get("text", ""))
    rows = read_any(data, payload.get("filename", ""))
    index, mapping = find_header_row(rows)
    assert_that(index >= 0, "Could not find a header row with a product name column.")
    created, skipped = [], []
    with db.tx():
        for row in rows[index + 1:]:
            def cell(field):
                position = mapping.get(field)
                if position is None or position >= len(row):
                    return ""
                return row[position]

            name = str(cell("name")).strip()
            if not name:
                continue
            gst_raw = to_number(cell("gst"))
            try:
                product = products.save(actor, {
                    "name": name,
                    "manufacturer": str(cell("mfr")).strip(),
                    "hsn": str(cell("hsn")).strip() or "3004",
                    "gst_rate": gst_raw if int(gst_raw) in GST_RATES else settings.num("default_gst_rate", 12),
                    "pack_size": parse_pack(cell("pack")) if "pack" in mapping else 1,
                })
                created.append(product["name"])
            except Exception as error:  # one bad row must not stop the import
                skipped.append({"name": name, "reason": str(error)})
    return {"created": len(created), "skipped": skipped, "names": created[:50]}
